http-webhook.md

```diff
@@ -0,0 +1,339 @@
+
+This specification also applies equivalently to HTTP/2 ([RFC7540][RFC7540]),
+which is compatible with HTTP 1.1 semantics.
+
+## 2. Delivering notifications
+
+### 2.1. Delivery request
+
+Notifications are delivered using a HTTP request. The response indicates the
+resulting status of the delivery.
+
+HTTP-over-TLS (HTTPS) [RFC2818][RFC2818] MUST be used for the connection.
+
+The HTTP method for the delivery request MUST be [POST][POST].
+
+The [`Content-Type`][Content-Type] header MUST be carried and the request MUST
+carry a notification payload of the given content type. Requests without
+payloads, e.g. where the notification is entirely expressed in HTTP headers,
+are not permitted by this specification.
+
+This specification does not further constrain the content of the notification,
+and it also does not prescribe the [HTTP target resource][RFC7230-Section-5-1]
+that is used for delivery.
+
+If the delivery target supports and requires [Abuse Protection][#4-abuse-protection],
+the delivery request MUST include the `Origin` header. The `Origin` header value
+is a DNS name expression that identifies the sending system.
+
+### 2.2. Delivery response
+
+The delivery response MAY contain a payload providing detail status
+information in the case of handling errors. This specification does not attempt
+to define such a payload.
+
+The response MUST NOT use any of the [3xx HTTP Redirect status codes][3xx] and
+the client MUST NOT follow any such redirection.
+
+If a delivery target has been retired, but the HTTP site still exists, the
+site SHOULD return a [410 "Gone"][410] status code and the sender SHOULD refrain
+from sending any further notifications.
+
+If the delivery target is unable to process the request due to exceeding a
+request rate limit, it SHOULD return a [429 Too Many Requests][429] status code
+and MUST include the [`Retry-After`][Retry-After] header. The sender MUST
+observe the value of the Retry-After header and refrain from sending further
+requests until the indicated time.
+
+If the delivery has been accepted and processed, and if the response carries a
+payload with processing details, the response MUST have the [200 OK][200] or
+[201 Created][201] status code. In this case, the response MUST carry a
+[`Content-Type`][Content-Type] header.
+
+If the delivery has been accepted and processed, but carries no payload, the
+response MUST have the [201 Created][201] or [204 No Content][204] status code.
+
+If the delivery has been accepted, but has not yet been processed or if the
+processing status is unknown, the response MUST have the [202 Accepted][202]
+status code.
+
+If the delivery cannot be accepted because the notification format has not
+been understood, the service MUST respond with status code [415][415].
+
+All further error status codes apply as specified in [RFC7231][RFC7231].
+
+## 3. Authorization
+
+The delivery request MUST use one of the following two methods, both of which
```

Comment:
Why is OAuth2 the only valid auth? E.g. GitHub uses content signatures in their webhooks. APNs uses connection-level auth by only allowing connections from known certificates.

Reply:
"The client MAY use any token-based authorization scheme. Challenge-based
schemes MUST NOT be used."

The point of this spec is not to document all variations of WebHooks presently in use, but rather define a straightforward model with a core set of options. You can form a token with a content signature and flow that in the Authorization header under its own scheme and be compliant. 

OAuth 2.0 provides a solid summary of how to do the token flow via Authorization header or via query string and that is why I lean on it.

The reason I'm not including client certificates is that they're notoriously terrible to manage. If/when the group thinks there is concrete need for client cert auth, then we should add it.

Reply:
For Adobe I/O Events, we use the client secret, not a client certificate, for producing a content signature. We already manage client secrets generated during clientID generation, so no added burden.

Reply:
I agree with @inlined and made the same comment on the workgroup call. GitHub uses HMAC / symmetric secret and this is adopted by other platforms also like Patreon.

> The delivery request MUST use one of the following two methods

What does this mean? Does it mean that if using auth we must implemented either - or that auth must be done and it must be either? @clemensv

Reply:
@alexellis  I added the following clause yesterday:

"The client MAY use any token-based authorization scheme. The token can take any shape, and can be a standardized token format or a simple key expression."

That means (and meant) using any key or an HMAC using a key is completely legitimate. I don't define or mandate a specific token format here. Azure Functions generates a symmetric key and requires that key for its default implementation that you get from "file new", and that would be conforming. What I need to mandate for interop is that the query argument has a well-known name and that's mostly what I do here.